ChunkedList: store pointers to next/last elements instead of indices …

michaelsippel opened the pull request:
…to save overhead of address calculation and allow more efficient packing of ItemAccess

* this also reintroduces the chunk-size as template parameter instead of storing it as member
* initialize AtomicList with `chunk_capacity`, not allocation size
* AtomicList: construct Item/ItemControlBlock from `memory::Block`
* ChunkedList: construct Chunk from `memory::Block`
* reorder ChunkedList::Item struct for more efficient alignment
* encode `has_item` of ItemAccess into one 8-byte pointer using bitmagic

Commit:
ChunkedList: store pointers to next/last elements instead of indices to save overhead of address calculation and allow more efficient packing of ItemAccess

* this also reintroduces the chunk-size as template parameter instead of storing it as member
* initialize AtomicList with `chunk_capacity`, not allocation size
* AtomicList: construct Item/ItemControlBlock from `memory::Block`
* ChunkedList: construct Chunk from `memory::Block`
* reorder ChunkedList::Item struct for more efficient alignment
* encode `has_item` of ItemAccess into one 8-byte pointer using bitmagic

michaelsippel commented:
Here I'm still getting warnings which I dont understand

```
warning: 'short unsigned int __atomic_fetch_sub_2(volatile void*, short unsigned int, int)' writing 2 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
  645 |       { return __atomic_fetch_sub(&_M_i, __i, int(__m)); }
```

psychocoderHPC commented:
> Here I'm still getting warnings which I dont understand
> 
> ```
> warning: 'short unsigned int __atomic_fetch_sub_2(volatile void*, short unsigned int, int)' writing 2 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
>   645 |       { return __atomic_fetch_sub(&_M_i, __i, int(__m)); }
> ```

Could it be a hint that some data are unaligned?

psychocoderHPC commented:
@michaelsippel Is this PR a replacement for #45?

psychocoderHPC commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -15,12 +15,13 @@
 #include <redGrapes/util/trace.hpp>
 #include <redGrapes/memory/allocator.hpp>
-#include <redGrapes/memory/bump_allocator.hpp>
+//#include <redGrapes/memory/bump_allocator.hpp>
```
?? why is this commented out?

michaelsippel replied:
commented because its not needed. Allocator is a template parameter. will be removed

psychocoderHPC replied:
please remove the line.

psychocoderHPC commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -172,46 +188,47 @@ struct ChunkedList
         {
-            for( unsigned i = 0; i < last_idx; ++i )
-                items()[i].~Item();
+            for( Item * item = first_item; item != last_item; item++ )
```
`last_item` is pointing by default to the element before the first item. If we never allocate memory but create a chunked_list does it mean we loop over non existing memory?

psychocoderHPC replied:
maybe `for( Item * item = first_item; item <= last_item ; item++ )`?

michaelsippel replied:
True, currently this bug  doesnt manifest because a chunk is only allocated if at least one element exists , but when we switch to the algorithm where the thread with the last slot allocates the new chunk, it can happen that a chunk remains empty and thus creating an endless loop.

psychocoderHPC commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -172,46 +188,47 @@ struct ChunkedList
         typename memory::AtomicList< Chunk, Allocator >::MutBackwardIterator chunk;
 
+        /* this pointer packs the address of the current element
+         * and the `has_element` bit in its MSB.
```
`MSB` ?? What is this?

michaelsippel replied:
MSB = most significant bit

psychocoderHPC replied:
please add this to the first usage

psychocoderHPC commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -441,7 +438,7 @@ struct ChunkedList
     {
         if( chunk->item_count.fetch_sub(1) == 0 )
-            chunks.erase( chunk );     
+            chunks.erase( chunk );
```
If a thread is entering this line and at the same moment another thread is pushing an item and is increasing the counter and sees that no item is free and then releases the memory again we will have a double free.

michaelsippel replied:
in this version we increment item_count only after we are safe to have gotten a slot.  Only if the increment of `item_count` happens first and then decrements this is relevant

psychocoderHPC commented on redGrapes/util/chunked_list.hpp:
```diff
@@ -453,20 +450,21 @@ struct ChunkedList
+                release_chunk(chunk);
             }
 
             auto prev_chunk = chunks.allocate_item();
```
How do we prevent that more than one thread is creating new memory?

psychocoderHPC reviewed (changes requested):
IMO we still have the double free issue I tried to solve with #45 and we have the problem that more than one thread can append a new chunk.

michaelsippel commented:
Yes you are right, this PR does not include the fix for the double allocation / memory leak yet ,  #45 is still neccesary, either one needs to be rebased

psychocoderHPC committed:
fix double free and double alloc

Equally to
https://github.com/ComputationalRadiationPhysics/redGrapes/pull/45 this
PR should solve the possible double free and double alloc.

psychocoderHPC commented:
I opened a PR against your PR to solve the double free and alloc issue: https://github.com/michaelsippel/redGrapes/pull/1
My local tests works well, please have a look.
As I wrote before it is not necessary that #45 get merged if this PR is solving the big issue with the double alloc and free.

Commit:
Merge pull request #1 from psychocoderHPC/fix-doubleFreeAndAlloc

fix double free and double alloc
